src/index.js

```diff
@@ -1,169 +1,323 @@
+	 */
+	maybeScrollToAnchor(hash, context) {
+		// Bail early if the hash is null
+		if (hash == null) {
+			return false;
+		}
+		const element = this.getAnchorElement(hash);
+		if (!element) {
+			console.warn(`Element ${hash} not found`);
+			return false;
+		}
+		if (!(element instanceof Element)) {
+			console.warn(`Element ${hash} is not a DOM node`);
+			return false;
+		}
+		const top =
+			element.getBoundingClientRect().top + window.pageYOffset - this.getOffset(element);
+		this.swup.scrollTo(top, this.shouldAnimate(context));
+		return true;
+	}
+
+	/**
+	 * Handles `transitionStart`
+	 * @param {PopStateEvent} popstate
+	 */
+	onTransitionStart = (popstate) => {
+		if (this.options.doScrollingRightAway && !this.swup.scrollToElement) {
+			this.doScrollingBetweenPages(popstate);
+		}
+	};
+
+	/**
+	 * Handles `contentReplaced`
+	 * @param {PopStateEvent} popstate
+	 */
+	onContentReplaced = (popstate) => {
+		if (!this.options.doScrollingRightAway || this.swup.scrollToElement) {
+			this.doScrollingBetweenPages(popstate);
+		}
+
+		this.restoreScrollContainers(popstate);
+	};
+
+	/**
+	 * Scrolls the window, based on context
+	 *
+	 * @param {PopStateEvent|boolean} popstate
+	 * @returns void
+	 */
+	doScrollingBetweenPages = (popstate) => {
+		const swup = this.swup;
+
+		// Bail early on popstate and inactive `animateHistoryBrowsing`
+		if (popstate && !swup.options.animateHistoryBrowsing) {
+			return;
+		}
+
+		// Try scrolling to a given anchor
+		if (this.maybeScrollToAnchor(swup.scrollToElement, 'betweenPages')) {
+			swup.scrollToElement = null;
+			return;
+		}
+
+		// Finally, scroll to either the stored scroll position or to the very top of the page
+		const scrollPositions = this.getStoredScrollPositions(getCurrentUrl());
+		const top = (scrollPositions.window && scrollPositions.window.top) || 0;
+		swup.scrollTo(top, this.shouldAnimate('betweenPages'));
+	};
+
+	/**
+	 * Stores the current scroll positions for the URL we just came from
+	 */
+	onWillReplaceContent = () => {
+		this.storeScrollPositions(this.previousUrl);
+		this.previousUrl = getCurrentUrl();
+	};
+
+	/**
+	 * Deletes the scroll positions for the URL a link is pointing to,
+	 * if shouldRestoreScrollPosition doesn't evaluate to true
+	 *
+	 * @param {event} e
+	 * @returns void
+	 */
+	onClickLink = (e) => {
+		if (!this.options.shouldRestoreScrollPosition(e.delegateTarget)) {
+			return;
+		}
+		const url = new Link(e.delegateTarget).getAddress();
+		this.deleteStoredScrollPositions(url);
+	};
+
+	/**
+	 * Stores the scroll positions for the current URL
+	 * @param {string} url
+	 */
+	storeScrollPositions(url) {
+		// create an object to store the scroll positions
+		const storeEntry = {
+			window: { top: window.scrollY, left: window.scrollX },
+			containers: []
+		};
+		// fill up the object with scroll positions for each matching element
+		const $containers = document.querySelectorAll(this.options.scrollContainers);
+		$containers.forEach((el) =>
```

Comment:
If we turn this into a `.map` and create the containers object directly, we'd save a few lines. Also maybe reorder to create the containers first, then construct the final entry?

(Untested)

```js
const window = { top: window.scrollY, left: window.scrollX };
const containers = queryAll(this.options.scrollContainers)
  .map((el) => ({ top: el.scrollTop, left: el.scrollLeft }));

this.scrollPositionsStore[url] = { window, containers };
```

Reply:
Nice, I'm old-school, keep forgetting that `.map` exists nowadays ;) ...`window` can't be used as a variable I think, but still – I will adopt your suggestion partly just to get used to thinking with `.map` more.

Reply:
Sure `window` can be assigned to a `const`; it's scoped locally to that method.

Reply:
ah sorry, wrong explanation – you just can't use it if you are immediately referencing window... Anyways, too complicated to explain, wait for the push :)

Reply:
Even then, no? You're grabbing the scroll values from the global window object, only then are you creating a new local const. Maybe I have it backwards, but I'm doing this all over the place and it seems to be working 🤡

Reply:
If I type this in the console:
```js
const window = { top: window.scrollY, left: window.scrollX };
```
You will get the error: 

```
Uncaught SyntaxError: Identifier 'window' has already been declared".
```

If I put it in the program, I get this error:

```
ReferenceError: Cannot access 'window2' before initialization
```

Reply:
Anyways, let's not make this more necessarily complicated than it has to be, shall we? ;)

Reply:
Indeed, you're right, the scope doesn't work out here. I'll stop spreading nonsense now 🧃

Reply:
🙏😁